File: src/main/java/Solution_2352_EqualRowAndColumnPairs.java
import java.util.*;

public class Solution_2352_EqualRowAndColumnPairs {
    public static void main(String[] args) {
        var grid = new int[][]{{3, 1, 2, 2}, {1, 4, 4, 4}, {2, 4, 2, 2}, {2, 5, 2, 2}};
        System.out.println(equalPairs(grid));
    }

    private static int equalPairs(int[][] grid) {
        var columns = new ArrayList<List<Integer>>();
        var rows = Arrays.stream(grid).map(row -> Arrays.stream(row).boxed().toList()).toList();

        for (int i = 0; i < grid.length; i++) { // столбец
            var column = new ArrayList<Integer>();
            for (int j = 0; j < grid.length; j++) { // строка
                column.add(grid[j][i]);
            }
            columns.add(column);
        }

        return (int) rows.stream()
                .flatMap(row -> columns.stream().filter(row::equals))
                .count();
    }

    public int equalPairs2(int[][] grid) {
        int size = grid.length;
        Map<String, Integer> pair = new HashMap<>();

        for (int[] row : grid) {
            String r = Arrays.toString(row);
            pair.put(r, pair.getOrDefault(r, 0) + 1);
        }

        int j = 0, count = 0;
        while (j < size) {
            int[] col = new int[size];
            for (int i = 0; i < size; i++) {
                col[i] = grid[i][j];
            }
            j++;
            count += pair.getOrDefault(Arrays.toString(col), 0);
        }
        return count;
    }
}

File: src/main/kotlin/Solution_2352_EqualRowAndColumnPairs.kt
/**
 * 2352 Equal Row and Column Pairs
 *
 * Medium
 *
 * Given a 0-indexed n x n integer matrix grid, return the number of pairs (ri, cj) such that row ri and column cj are equal.
 *
 * A row and column pair is considered equal if they contain the same elements in the same order (i.e., an equal array).
 *
 *
 *
 * Example 1:
 *
 *
 * Input: grid = [[3,2,1],[1,7,6],[2,7,7]]
 * Output: 1
 * Explanation: There is 1 equal row and column pair:
 * - (Row 2, Column 1): [2,7,7]
 * Example 2:
 *
 *
 * Input: grid = [[3,1,2,2],[1,4,4,5],[2,4,2,2],[2,4,2,2]]
 * Output: 3
 * Explanation: There are 3 equal row and column pairs:
 * - (Row 0, Column 0): [3,1,2,2]
 * - (Row 2, Column 2): [2,4,2,2]
 * - (Row 3, Column 2): [2,4,2,2]
 */
fun main() {
//    val grid = arrayOf(intArrayOf(3, 2, 1), intArrayOf(1, 7, 6), intArrayOf(2, 7, 7))
    val grid = arrayOf(intArrayOf(3, 1, 2, 2), intArrayOf(1, 4, 4, 4), intArrayOf(2, 4, 2, 2), intArrayOf(2, 5, 2, 2))
    println(equalPairs(grid))
}

fun equalPairs(grid: Array<IntArray>): Int {
    // массив для хранения колонок
    val columns = Array(grid.size) { List(0) { 0 } }
    // массив строк
    val rows = grid.map { it.toList() }

    // заполняем колонки
    for (i in grid.indices) {
        val column = mutableListOf<Int>()
        for (j in grid.indices) {
            column.add(grid[j][i])
        }
        columns[i] = column
    }

    // считаем количество совпадений колонок и строк
    var resCount = 0
    for (row in rows) {
        val countSameColumns = columns.count { it == row }
        resCount += countSameColumns
    }

    return resCount
}